### PR TITLE
PWN-9575 - New send service integration

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-9575

## Description of Work

First iteration for New Send Service integration

Taking root changes from https://github.com/p2p-org/key-app-android/pull/1745

## Screenshots (Optional)

| Screenshot #1 | Screenshot #2 | Screenshot #3 |
|:-------------:|:-------------:|:-------------:|
| [Paste Pic 1] | [Paste Pic 2] | [Paste Pic 3] |

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/newsend/SendModule.kt [hunk 1]
```diff
@@ -90,5 +99,17 @@ object SendModule : InjectionModule {
         factoryOf(::FeeRelayerViaLinkInteractor)
         factoryOf(::SendViaLinkInteractor)
         factoryOf(::UserSendLinksDatabaseRepository) bind UserSendLinksLocalRepository::class
+
+        factoryOf(::SendServiceInteractor)
+        singleOf(::SendServiceRemoteRepository)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/SendModule.kt:104]: you sure you want to make remote repo as a singleton? and without interface?

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateSendTransactionRequest.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.google.gson.reflect.TypeToken
+import java.lang.reflect.Type
+import org.p2p.core.rpc.JsonRpc
+
+class GenerateSendTransactionRequest(
+    val userWallet: String,
+    val mintAddress: String,
+    val amount: String,
+    val recipient: String
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateSendTransactionRequest.kt:11]: try using SolAddress or Base58String

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package org.p2p.wallet.newsend.api
+
+import com.google.gson.annotations.SerializedName
+import java.math.BigDecimal
+import java.math.BigInteger
+
+data class GenerateTransactionResponse(
+    @SerializedName("transaction")
+    val transaction: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:9]: base58String / base64string
making string format here is important

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @SerializedName("blockhash")
+    val blockhash: String,
+    @SerializedName("expiresAt")
+    val expiresAt: Long,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:13]: MillisSinceEpoch

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package org.p2p.wallet.newsend.api
+
+import com.google.gson.annotations.SerializedName
+import java.math.BigDecimal
+import java.math.BigInteger
+
+data class GenerateTransactionResponse(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:7]: GenerateSendTransactionResponse
we can generate transaction for a lot of things

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    val networkFee: NetworkFeeResponse,
+    @SerializedName("token_account_rent")
+    val tokenAccountRent: TokenAccountRentResponse,
+)
+
+data class NetworkFeeResponse(
+    @SerializedName("source")
+    val source: String,
+    @SerializedName("amount")
+    val usd_amount: AmountResponse
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:30]: usdAmount

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+data class TokenAccountRentResponse(
+    @SerializedName("source")
+    val source: String,
+    @SerializedName("amount")
+    val usd_amount: AmountResponse
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:37]: usdAmount

File: app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+data class AmountResponse(
+    @SerializedName("amount")
+    val amount: BigInteger,
+    @SerializedName("usd_amount")
+    val usd_amount: BigDecimal,
+    @SerializedName("mint")
+    val mint: String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/api/GenerateTransactionResponse.kt:46]: tokenMint

File: app/src/main/java/org/p2p/wallet/newsend/model/SearchResult.kt [hunk 9]
```diff
@@ -58,3 +58,8 @@ sealed class SearchResult(open val address: String) : Parcelable {
 
 fun TemporaryAccount.toSearchResult(): SearchResult =
     SearchResult.AddressFound(address)
+
+fun SearchResult.nicknameOrAddress(): String {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/model/SearchResult.kt:62]: usernameOrAddress
we don't use nickname domain

File: app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRemoteRepository.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import org.p2p.wallet.newsend.api.GenerateSendTransactionRequest
+import org.p2p.wallet.newsend.api.GenerateTransactionResponse
+
+class SendServiceRemoteRepository(
+    private val api: RpcApi,
+    private val gson: Gson,
+    urlProvider: NetworkServicesUrlProvider
+) : SendServiceRepository {
+
+    private val sendServiceStringUrl = urlProvider.loadSendServiceEnvironment().baseServiceUrl.toURI()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRemoteRepository.kt:18]: won't be changed if we change it in debug settings
try making it as get()

File: app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRemoteRepository.kt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    override suspend fun generateTransaction(
+        userPublicKey: Base64String,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRemoteRepository.kt:21]: userPublicKey is Base58String, so as tokenMint and recipientAddress

File: app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRepository.kt [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package org.p2p.wallet.newsend.repository
+
+import java.math.BigInteger
+import org.p2p.core.crypto.Base64String
+import org.p2p.wallet.newsend.api.GenerateTransactionResponse
+
+interface SendServiceRepository {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRepository.kt:7]: is not used in DI + should return domain entites

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/newsend/ui/main/TAG.kt
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/newsend/ui/main/TAG.kt:1]: TAG ?

File: app/src/main/java/org/p2p/wallet/newsend/SendModule.kt (hunk 1 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/newsend/SendModule.kt:104]: agree, there has `SendServiceRepository` interface and stateless `SendServiceRemoteRepository`, so it easily can be a `factoryOf(::SendServiceRemoteRepository) bind SendServiceRepository::class`

File: app/src/main/java/org/p2p/wallet/newsend/model/SendInitialData.kt [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        if (userTokens.isEmpty()) {
+            throw SendFatalError("User non-zero tokens can't be empty!")
+        }
+
+        return userTokens.first()
+    }
+
+    fun findSolToken(sourceToken: Token.Active, solToken: Token.Active?): Token.Active {
```
Comment (eduardmaximovich) [inline app/src/main/java/org/p2p/wallet/newsend/model/SendInitialData.kt:37]: To make it clearer:
```suggestion
    @Throws(SendFatalError::class)
    fun getSolTokenOrThrow(sourceToken: Token.Active, solToken: Token.Active?): Token.Active {
```

**Review by eduardmaximovich — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRepository.kt (hunk 12 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/newsend/repository/SendServiceRepository.kt:7]: yeap, not ready yet. 
Will do in a separate PR

File: app/src/main/java/org/p2p/wallet/newsend/ui/main/TAG.kt
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/newsend/ui/main/TAG.kt:1]: Just moving files issues.

**Comment by gslevinkov** [comment]
closed due to redundancy

**Pull request closed without merging**